File: app/src/test/java/com/mavenkalabs/adskipper/util/ConfigReaderTest.java
package com.mavenkalabs.adskipper.util;

import org.junit.Assert;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class ConfigReaderTest {

    @Test()
    public void verifyConfigRetrieval() throws Exception {
        final AtomicBoolean callbackInvoked = new AtomicBoolean();
        CountDownLatch latch = new CountDownLatch(1);
        try (final ConfigReader ignored = new ConfigReader(config -> {
            callbackInvoked.set(true);
            latch.countDown();
        })){
            if (latch.await(2, TimeUnit.SECONDS)) {
                Assert.assertTrue(callbackInvoked.get());
            } else {
                fail("callback was not invoked");
            }
        }
    }


    @Test()
    public void verifyConfigNotRetrievedWhenNoChange() throws Exception {
        final AtomicInteger callbackCount = new AtomicInteger();
        CountDownLatch latch = new CountDownLatch(1);
        try (ConfigReader configReader = new ConfigReader(config -> {
            callbackCount.incrementAndGet();
            latch.countDown();
        })){
            if (latch.await(2, TimeUnit.SECONDS)) {
                configReader.loadConfig();
                assertEquals(1, callbackCount.get());
            } else {
                fail("callback was not invoked");
            }

        }
    }
}

File: app/src/test/kotlin/com/mavenkalabs/adskipper/util/ConfigReaderTest.kt
package com.mavenkalabs.adskipper.util

import junit.framework.TestCase.assertEquals
import org.junit.Assert
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlin.test.fail

internal class ConfigReaderTest {

    @Test
    fun verifyConfigRetrieval() {
        val callbackInvoked = AtomicBoolean(false)
        val latch = CountDownLatch(1)
        ConfigReader {
            callbackInvoked.set(true)
            latch.countDown()
        }.use {
            if (latch.await(2, TimeUnit.SECONDS)) {
                Assert.assertTrue(callbackInvoked.get())
            } else {
                fail("callback was not invoked")
            }
        }

    }

    @Test
    fun verifyConfigNotRetrievedWhenNoChange() {
        val callbackCount = AtomicInteger(0)
        val latch = CountDownLatch(1)
        ConfigReader {
                callbackCount.incrementAndGet()
                latch.countDown()
        }.use { configReader ->
            {
                if (latch.await(2, TimeUnit.SECONDS)) {
                    configReader.loadConfig()
                    assertEquals(1, callbackCount.get())
                } else {
                    fail("callback was not invoked")
                }
            }
        }
    }
}
